distributed/protocol/torch.py

```diff
@@ -7,7 +7,12 @@
 @dask_serialize.register(torch.Tensor)
 def serialize_torch_Tensor(t):
     requires_grad_ = t.requires_grad
-    header, frames = serialize(t.detach_().numpy())
+
+    try:
+        header, frames = serialize(t.detach_().numpy())
```

Comment:
I don't think we want to use `detach_` here because it modifies `t` in place:

``` python
>>> import torch
>>> x = torch.rand(4, requires_grad=True)
>>> y = x.detach_().numpy()
>>> x.requires_grad
False
```

A test should be added to make sure that `x.requires_grad` stays the same before/after serialization.

Reply:
I originally placed a `.detach()` but after some discussions, we decided to add `.detach_()` instead. The reason you said, and it made sense for me, was that `.detach_() ` creates a leaf and does not break the graph.  What we want is just to extract the data from the tensor object, and if it requires or not gradients, to be de/serialized correctly. This operation shouldn't perturb what the model expects from these tensors. Are we sure we have to get back to a plain `.detach()`. 

This would be the commit to address what's been reviewed here:

```
diff --git a/distributed/protocol/tests/test_torch.py b/distributed/protocol/tests/test_torch.py
index e6747e32..fa181a0d 100644
--- a/distributed/protocol/tests/test_torch.py
+++ b/distributed/protocol/tests/test_torch.py
@@ -21,8 +21,17 @@ def test_grad():
     t.requires_grad = True
 
     t2 = deserialize(*serialize(t))
-    assert (t2.detach_().numpy() == x).all()
+
+    assert (t2.detach().numpy() == x).all()
     assert (t2.grad.numpy() == 1).all()
+    assert (t2.requires_grad is True)
+
+    t.requires_grad = False
+    t3 = deserialize(*serialize(t))
+
+    assert (t3.detach().numpy() == x).all()
+    assert (t3.grad.numpy() == 1).all()
+    assert (t3.requires_grad is False)
 
 
 def test_resnet():
diff --git a/distributed/protocol/torch.py b/distributed/protocol/torch.py
index e4185468..a70e60b6 100644
--- a/distributed/protocol/torch.py
+++ b/distributed/protocol/torch.py
@@ -8,9 +8,9 @@ import numpy as np
 def serialize_torch_Tensor(t):
     requires_grad_ = t.requires_grad
 
-    try:
+    if requires_grad_:
         header, frames = serialize(t.detach_().numpy())
-    except RuntimeError:
+    else:
         header, frames = serialize(t.numpy())
 
     if t.grad is not None:

```

Thoughts :)?

Reply:
> .detach_() creates a leaf and does not break the graph.

I think this is true for `detach()` too. I think https://github.com/dask/distributed/pull/2586#issuecomment-478116967 is concerned with correctly sending the `requires_grad` information.

Serialization should not implicitly modify the inputs.

Reply:
I changed it back to `detach()`.